### ISSUE TITLE
Failing to build `torch-sys-0.1.17` on Ubuntu (16.04) (CUDA)

### ISSUE THREAD

**Issue by kahlil29** [LaurentMazare/tch-rs#222]
I have installed Anaconda and then built `libtorch` (1.15.0) from source using the instructions provided [here](https://github.com/pytorch/pytorch#installation). 
But now my rust project encounters the following error while building : 

```
error: failed to run custom build command for `torch-sys v0.1.7`

Caused by:
  process didn't exit successfully: `/home/ubuntu/rust-foodScoring-tf-inference/target/debug/build/torch-sys-9793dd7018dbafc8/build-script-build` (exit code: 1)
--- stdout
cargo:rustc-link-search=native=/home/ubuntu/anaconda3/lib
TARGET = Some("x86_64-unknown-linux-gnu")
OPT_LEVEL = Some("0")
HOST = Some("x86_64-unknown-linux-gnu")
CXX_x86_64-unknown-linux-gnu = None
CXX_x86_64_unknown_linux_gnu = None
HOST_CXX = None
CXX = None
CXXFLAGS_x86_64-unknown-linux-gnu = None
CXXFLAGS_x86_64_unknown_linux_gnu = None
HOST_CXXFLAGS = None
CXXFLAGS = None
CRATE_CC_NO_DEFAULTS = None
DEBUG = Some("true")
CARGO_CFG_TARGET_FEATURE = Some("fxsr,sse,sse2")
running: "c++" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-m64" "-I" "/home/ubuntu/anaconda3/include" "-I" "/home/ubuntu/anaconda3/include/torch/csrc/api/include" "-Wl,-rpath=/home/ubuntu/anaconda3/lib" "-std=c++14" "-D_GLIBCXX_USE_CXX11_ABI=1" "-o" "/home/ubuntu/rust-foodScoring-tf-inference/target/debug/build/torch-sys-448e51ef5f47d1fc/out/libtch/torch_api.o" "-c" "libtch/torch_api.cpp"
cargo:warning=libtch/torch_api.cpp:1:39: fatal error: torch/csrc/autograd/engine.h: No such file or directory
cargo:warning=compilation terminated.
exit code: 1
running: "c++" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-m64" "-I" "/home/ubuntu/anaconda3/include" "-I" "/home/ubuntu/anaconda3/include/torch/csrc/api/include" "-Wl,-rpath=/home/ubuntu/anaconda3/lib" "-std=c++14" "-D_GLIBCXX_USE_CXX11_ABI=1" "-o" "/home/ubuntu/rust-foodScoring-tf-inference/target/debug/build/torch-sys-448e51ef5f47d1fc/out/libtch/fake_cuda_dependency.o" "-c" "libtch/fake_cuda_dependency.cpp"
exit code: 0

--- stderr
/home/ubuntu/rust-foodScoring-tf-inference/target/debug/build/torch-sys-9793dd7018dbafc8/build-script-build: /home/ubuntu/anaconda3/lib/libcurl.so.4: no version information available (required by /home/ubuntu/rust-foodScoring-tf-inference/target/debug/build/torch-sys-9793dd7018dbafc8/build-script-build)


error occurred: Command "c++" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-m64" "-I" "/home/ubuntu/anaconda3/include" "-I" "/home/ubuntu/anaconda3/include/torch/csrc/api/include" "-Wl,-rpath=/home/ubuntu/anaconda3/lib" "-std=c++14" "-D_GLIBCXX_USE_CXX11_ABI=1" "-o" "/home/ubuntu/rust-foodScoring-tf-inference/target/debug/build/torch-sys-448e51ef5f47d1fc/out/libtch/torch_api.o" "-c" "libtch/torch_api.cpp" with args "c++" did not execute successfully (status code exit code: 1).



warning: build failed, waiting for other jobs to finish...
error: build failed
```
Any idea what the issue is?

**Comment by LaurentMazare** [comment]
Did you follow the steps from the [readme](https://github.com/LaurentMazare/tch-rs#libtorch-manual-install) and set LIBTORCH to where the libtorch directory that you're building is?
Looking at the error message, some libtorch header cannot be found so it may be that this LIBTORCH is not set properly as it tells cargo/the c++ compiler where to find the header.

**Comment by kahlil29** [comment]
@LaurentMazare thanks for the swift reply. 
Yes I suspected that so I double checked and it turns out there was a path error. Now I located the `libtorch_cuda.so` file along with others. So it turns out they're in a `build` (and then subdirectory `lib`) folder under the `pytorch` folder (pytorch repo cloned from GitHub to build from source). 
So I've correct it and updated my `LD_LIBRARY_PATH` to include `/home/ubuntu/source_repos/pytorch/build/lib` 
I did a `cargo clean` and attempted the `cargo run` again
I still get a similar error : 

```
error: failed to run custom build command for `torch-sys v0.1.7`

Caused by:
  process didn't exit successfully: `/home/ubuntu/rust-foodScoring-tf-inference/target/debug/build/torch-sys-9793dd7018dbafc8/build-script-build` (exit code: 1)
--- stdout
cargo:rustc-link-search=native=/home/ubuntu/source_repos/pytorch/build/lib
TARGET = Some("x86_64-unknown-linux-gnu")
OPT_LEVEL = Some("0")
HOST = Some("x86_64-unknown-linux-gnu")
CXX_x86_64-unknown-linux-gnu = None
CXX_x86_64_unknown_linux_gnu = None
HOST_CXX = None
CXX = None
CXXFLAGS_x86_64-unknown-linux-gnu = None
CXXFLAGS_x86_64_unknown_linux_gnu = None
HOST_CXXFLAGS = None
CXXFLAGS = None
CRATE_CC_NO_DEFAULTS = None
DEBUG = Some("true")
CARGO_CFG_TARGET_FEATURE = Some("fxsr,sse,sse2")
running: "c++" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-m64" "-I" "/home/ubuntu/source_repos/pytorch/build/include" "-I" "/home/ubuntu/source_repos/pytorch/build/include/torch/csrc/api/include" "-Wl,-rpath=/home/ubuntu/source_repos/pytorch/build/lib" "-std=c++14" "-D_GLIBCXX_USE_CXX11_ABI=1" "-o" "/home/ubuntu/rust-foodScoring-tf-inference/target/debug/build/torch-sys-448e51ef5f47d1fc/out/libtch/torch_api.o" "-c" "libtch/torch_api.cpp"
cargo:warning=libtch/torch_api.cpp:1:39: fatal error: torch/csrc/autograd/engine.h: No such file or directory
cargo:warning=compilation terminated.
exit code: 1
running: "c++" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-m64" "-I" "/home/ubuntu/source_repos/pytorch/build/include" "-I" "/home/ubuntu/source_repos/pytorch/build/include/torch/csrc/api/include" "-Wl,-rpath=/home/ubuntu/source_repos/pytorch/build/lib" "-std=c++14" "-D_GLIBCXX_USE_CXX11_ABI=1" "-o" "/home/ubuntu/rust-foodScoring-tf-inference/target/debug/build/torch-sys-448e51ef5f47d1fc/out/libtch/dummy_cuda_dependency.o" "-c" "libtch/dummy_cuda_dependency.cpp"
exit code: 0

--- stderr


error occurred: Command "c++" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-m64" "-I" "/home/ubuntu/source_repos/pytorch/build/include" "-I" "/home/ubuntu/source_repos/pytorch/build/include/torch/csrc/api/include" "-Wl,-rpath=/home/ubuntu/source_repos/pytorch/build/lib" "-std=c++14" "-D_GLIBCXX_USE_CXX11_ABI=1" "-o" "/home/ubuntu/rust-foodScoring-tf-inference/target/debug/build/torch-sys-448e51ef5f47d1fc/out/libtch/torch_api.o" "-c" "libtch/torch_api.cpp" with args "c++" did not execute successfully (status code exit code: 1).

```

It seems to be referring to paths that don't exist like `/home/ubuntu/source_repos/pytorch/build/include/torch/csrc/api/include` 
I checked the `.../pytorch/build/include/` folder and it only contains : `c10d  sleef.h`

**Comment by LaurentMazare** [comment]
I'm not sure what this `build` directory is. When compiling from source, I always got `lib` in my `pytorch` directory.
Does `pytorch/include` exists and contains the file that the compiler complains about ? (this would be something like `pytorch/include/torch/csrc/autograd/engine.h`)
If that's the case maybe you just want to set `LIBTORCH` to the actual `pytorch` directory and move `lib` outside of `build`.

**Comment by kahlil29** [comment]
Okay I found the files. So the correct directory was `.../pytorch/torch` and that contains all the stuff needed. It built successfully now. Thanks for the help. 

There's another question I wanted to ask you but instead of opening another ticket (unless you would prefer that) : 
Have you tried or know about running model inference in parallel using `tch-rs`? I have around 4 models and the total time taken to run inference for a Tensor (of a single image) is almost 400ms. 
Do you think using `threads` via std library or `crossbeam` or using `rayon` would help?

**Comment by LaurentMazare** [comment]
Glad that it ended up working.
I haven't run multiple inferences in parallel. If the model tensors are all disjoints that should be fine though. Speed improvement would certainly depend on the percentage of the gpu that your models get to use, if each of them is able to use most of the gpu, there probably won't be much of a speedup. Using `threads` or `rayon` sounds reasonable (I don't have experience with `crossbeam`). Note that I never tried it so it may be the case that the pytorch api is not thread-safe and you'll run into some segfaults/weird behavior.

**Comment by kahlil29** [comment]
Thanks a lot, I'll probably try this and get back with some results by next week :)
